storage/src/adb/operation.rs

```diff
@@ -68,89 +61,48 @@ impl<K: Array, V: Array> Operation<K, V> {
-
-        Self {
-            data,
-            _phantom: std::marker::PhantomData,
-        }
-    }
-
-    /// Create a new commit operation that indicates the current floor on inactive operations is
-    /// `loc`.
-    pub fn commit(loc: u64) -> Self {
-        let mut data = Vec::with_capacity(Self::SIZE);
-        data.push(Self::COMMIT_CONTEXT);
-        data.extend_from_slice(&loc.to_be_bytes());
-        data.resize(Self::SIZE, 0);
-
-        Self {
-            data,
-            _phantom: std::marker::PhantomData,
-        }
-    }
-
-    pub fn to_key(&self) -> K {
-        K::decode(&self.data[1..K::SIZE + 1]).unwrap()
-    }
-
-    pub fn to_type(&self) -> Type<K, V> {
-        let key = K::decode(&self.data[1..K::SIZE + 1]).unwrap();
-        match self.data[0] {
-            Self::DELETE_CONTEXT => Type::Deleted(key),
-            Self::UPDATE_CONTEXT => {
-                let value = V::decode(&self.data[K::SIZE + 1..]).unwrap();
-                Type::Update(key, value)
-            }
-            Self::COMMIT_CONTEXT => {
-                let loc = u64::from_be_bytes(self.data[1..9].try_into().unwrap());
-                Type::Commit(loc)
-            }
-            _ => unreachable!(),
+    /// If this is a [Operation::Update] or [Operation::Deleted] operation, returns the key.
+    /// Otherwise, returns None.
+    pub fn to_key(&self) -> Option<K> {
+        match self {
+            Operation::Deleted(key) => Some(key.clone()),
+            Operation::Update(key, _) => Some(key.clone()),
+            Operation::Commit(_) => None,
         }
     }
 
+    ///If this is a [Operation::Update] operation, returns the value.
+    /// Otherwise, returns None.
     pub fn to_value(&self) -> Option<V> {
-        match self.data[0] {
-            Self::DELETE_CONTEXT => None,
-            Self::UPDATE_CONTEXT => Some(V::decode(&self.data[K::SIZE + 1..]).unwrap()),
-            Self::COMMIT_CONTEXT => None,
-            _ => unreachable!(),
+        match self {
+            Operation::Deleted(_) => None,
+            Operation::Update(_, value) => Some(value.clone()),
+            Operation::Commit(_) => None,
         }
     }
 }
 
 impl<K: Array, V: Array> Write for Operation<K, V> {
     fn write(&self, buf: &mut impl BufMut) {
-        assert!(self.data.len() == Self::SIZE);
-        buf.put_slice(&self.data);
+        match &self {
+            Operation::Deleted(k) => {
+                buf.put_u8(Self::DELETE_CONTEXT);
+                k.write(buf);
+                // Put 0s for the value
+                buf.put_bytes(0, V::SIZE);
+            }
+            Operation::Update(k, v) => {
+                buf.put_u8(Self::UPDATE_CONTEXT);
+                k.write(buf);
+                v.write(buf);
+            }
+            Operation::Commit(loc) => {
+                buf.put_u8(Self::COMMIT_CONTEXT);
+                buf.put_slice(&loc.to_be_bytes());
+                // Put 0s for the value
```

Comment:
nit: Put 0s for the rest of the key and the value (?)

Reply:
Updated comment to `// Pad with 0 up to [Self::SIZE]`